db/src/db.rs

```diff
@@ -899,135 +970,216 @@ mod tests {
+        assert_matches!(conn.last_transaction(),
+                        "[[100 :db.schema/attribute 100 ?tx true]
+                          [101 :db.schema/attribute 101 ?tx true]
+                          [?tx :db/txInstant ?ms ?tx true]]");
+        assert_matches!(conn.datoms(),
+                        "[[100 :db/ident :name/Ivan]
+                          [100 :db.schema/attribute 100]
+                          [101 :db/ident :name/Petr]
+                          [101 :db.schema/attribute 101]]");
+
+        // Upserting a tempid works.  The ref doesn't have to exist (at this time), but we can't
+        // reuse an existing ref due to :db/unique :db.unique/value.
+        let _report = conn.transact("[[:db/add \"t1\" :db/ident :name/Ivan]
+                                      [:db/add \"t1\" :db.schema/attribute 102]]").unwrap();
+        assert_matches!(conn.last_transaction(),
+                        "[[100 :db.schema/attribute 102 ?tx true]
+                          [?tx :db/txInstant ?ms ?tx true]]");
+        assert_matches!(conn.datoms(),
+                        "[[100 :db/ident :name/Ivan]
+                          [100 :db.schema/attribute 100]
+                          [100 :db.schema/attribute 102]
+                          [101 :db/ident :name/Petr]
+                          [101 :db.schema/attribute 101]]");
+
+        // A single complex upsert allocates a new entid.
+        let _report = conn.transact("[[:db/add \"t1\" :db.schema/attribute \"t2\"]]").unwrap();
+        assert_matches!(conn.last_transaction(),
+                        "[[65536 :db.schema/attribute 65537 ?tx true]
+                          [?tx :db/txInstant ?ms ?tx true]]");
+
+        // Conflicting upserts fail.
+        let err = conn.transact("[[:db/add \"t1\" :db/ident :name/Ivan]
+                                  [:db/add \"t1\" :db/ident :name/Petr]]").unwrap_err().to_string();
+        assert_eq!(err, "not yet implemented: Conflicting upsert: tempid \'t1\' resolves to more than one entid: 100, 101");
+
+        // tempids in :db/retract that don't upsert fail.
+        let err = conn.transact("[[:db/retract \"t1\" :db/ident :name/Anonymous]]").unwrap_err().to_string();
+        assert_eq!(err, "not yet implemented: [:db/retract ...] entity referenced tempid that did not upsert: t1");
+
+        // tempids in :db/retract that do upsert are retracted.  The ref given doesn't exist, so the
+        // assertion will be ignored.
+        let _report = conn.transact("[[:db/add \"t1\" :db/ident :name/Ivan]
+                                      [:db/retract \"t1\" :db.schema/attribute 103]]");
+        assert_matches!(conn.last_transaction(),
+                        "[[?tx :db/txInstant ?ms ?tx true]]");
+
+        // A multistep upsert.  The upsert algorithm will first try to resolve "t1", fail, and then
+        // allocate both "t1" and "t2".
+        let _report = conn.transact("[[:db/add \"t1\" :db/ident :name/Josef]
```

Comment:
One thing I see and will fix throughout is that Rust thinks that `let _...` is a usage.  In this case, I forgot to `.unwrap()`; if there was no `let _...`, the compiler would tell me I forgot to use the result.